### PR TITLE
Fix gzip and caching tools integration (#1190)

### LINKED ISSUE #1190: Static files served by 'tools.staticdir' cause decompression errors in browsers when both 'tools.gzip' and 'tools.caching' are on, but only when served from the cache.
Originally reported by: **Michiel Overtoom (Bitbucket: [MichielOvertoom](http://bitbucket.org/MichielOvertoom), GitHub: @motoom)**

---

I think I found a bug when both 'tools.gzip' and 'tools.caching' are on, for static files served by 'tools.staticdir'. Static content is placed uncompressed in the cache, but incorrectly marked as compressed, and the
result is that only the first serving is send correctly compressed to the client. Subsequent requests cause errors in the browser.

By the way, no exceptions are thrown by CherryPy.

I have attached a small test program which demonstrates this problem. Reloading the static file in the browser should produce the problem.

---
- Bitbucket: https://bitbucket.org/cherrypy/cherrypy/issue/1190

### PR DESCRIPTION
**Pull request by hexaclock**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Fixes an issue where an initial page load (when having both gzip and caching enabled) may cause a content decode error. Subsequent page loads work correctly.

* **What is the related issue number (starting with `#`)**

#1190

* **What is the current behavior?** (You can also link to an open issue here)

https://github.com/cherrypy/cherrypy/issues/1190

* **What is the new behavior (if this is a feature change)?**

* **Other information**:

### CODE REVIEW

File: cherrypy/lib/caching.py [hunk 1]
```diff
@@ -411,7 +411,9 @@ def tee(body):
                              body, response.time), len(body))
 
     response = cherrypy.serving.response
-    response.body = tee(response.body)
+    tee_gen = tee(response.body)
+    if tee_gen:
```
Comment (webknjaz) [inline cherrypy/lib/caching.py]: could you please explain your solution in more detail (and maybe comment/docstring in code)?

**Comment by webknjaz** [comment]
@hexaclock thanks for your contribution!
Any ideas about how to test this? If not, I'll need more time to analyse this change and think about the consequences ;)

**Comment by hexaclock** [comment]
It forces the CachingTool to now run after the encoding tool. In turn, this ensures the `tee_output` attached at before_finalize runs after the content is already gzipped as per `_d.gzip = Tool('before_finalize', encoding.gzip, priority=80)`

**Comment by webknjaz** [comment]
Sounds fair, now it's important to check whether this ordering change does not interfere with other tools. I'll also think about testing as suggested in #1190.

**Comment by hexaclock** [comment]
@webknjaz any luck? I may have some time this weekend to take a stab at writing a test for this based on @motoom's example in #1190.

**Comment by webknjaz** [comment]
@hexaclock sorry, I didn't have time for this yet.

**Comment by hexaclock** [comment]
@webknjaz 
I tried my test against the old _cptools.py and my edits - fails on the old and works on the new, as expected. Modeled after @motoom's paste in #1190.

File: cherrypy/test/test_caching.py [hunk 2]
```diff
@@ -272,6 +286,40 @@ def testExpiresTool(self):
...
+        
+        self.getPage(idx_uri, method='GET', headers=headers)
+        idx_resp_headers = dict(self.headers)
+        idx_gz_content_len = idx_resp_headers['Content-Length']
+        
+        self.getPage(jpg_uri, method='GET', headers=headers)
+        jpg_resp_headers = dict(self.headers)
+        jpg_gz_content_len = jpg_resp_headers['Content-Length']
+        
+        for _ in range(3):
```
Comment (webknjaz) [inline cherrypy/test/test_caching.py]: Is if required to have 3 queries? Wouldn't it be enough to have 2?

File: cherrypy/test/test_caching.py [hunk 3]
```diff
@@ -129,9 +129,23 @@ def wrongtype(self):
...
+            'tools.gzip.mime_types': ['text/*', 'image/*'],
+            'tools.caching.on': True,
+            'tools.staticdir.on': True,
+            'tools.staticdir.dir': 'static',
+            'tools.staticdir.root': curdir
+            
+        })
+        class GzipStaticCache(object):
+            @cherrypy.expose
+            def dummy(self):
```
Comment (webknjaz) [inline cherrypy/test/test_caching.py]: I believe, this handler not needed.

File: cherrypy/test/test_caching.py (hunk 2 above)
Comment (hexaclock) [reply cherrypy/test/test_caching.py]: Ah yes, you're right. Updated in latest commit.

File: cherrypy/test/test_caching.py (hunk 3 above)
Comment (hexaclock) [reply cherrypy/test/test_caching.py]: Removed.

**Comment by hexaclock** [comment]
@webknjaz Assuming this builds ok, are we good to go?

**Comment by hexaclock** [comment]
Seems to have failed due to what seems to be an environment problem: https://pastebin.com/vbhq7kXn

**Comment by webknjaz** [comment]
@hexaclock I've restarted failed jobs

**Comment by webknjaz** [comment]
Overall, looks good, but I still feel a need to check other tools priorities. I don't want this to affect other stuff.

**Comment by webknjaz** [comment]
Okay, I've added some test refactoring and it should be ok now to merge it after tests complete.

**Pull request merged**